Mathlib/AlgebraicGeometry/Curve/Smooth.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2024 Fangming Li. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fangming Li
+-/
+import Mathlib.AlgebraicGeometry.Scheme
+import Mathlib.RingTheory.Smooth.Basic
+
+/-!
+# Smooth morphisms of schemes
+
+In this file, we define when a morphism of schemes is smooth.
+-/
+
+/--
+Smooth homomorphisms between rings.
+-/
+def RingHom.IsSmooth {R A : Type _} [CommRing R] [CommRing A] (f : R →+* A) : Prop :=
+  @Algebra.Smooth R _ A _ (RingHom.toAlgebra f)
+
+namespace AlgebraicGeometry
+
+section ClosedImmersion
+
+/--
+Closed immersions between locally ringed spaces.
+-/
+class LocallyRingedSpace.IsClosedImmersion {X : LocallyRingedSpace} {Y : LocallyRingedSpace}
+    (f : X ⟶ Y) : Prop where
+  val_base_closed : ClosedEmbedding f.val.base
+  epi : ∀ (x : Y.toPresheafedSpace),
+    Function.Surjective ((TopCat.Presheaf.stalkFunctor CommRingCat x).map f.val.c)
+
+/--
+Closed immersions between locally ringed spaces.
+-/
+abbrev IsClosedImmersion {X Y : Scheme} (f : X ⟶ Y) : Prop :=
+  LocallyRingedSpace.IsClosedImmersion f
+
+end ClosedImmersion
+
+section Smooth
+
+/--
+Given a scheme morphism `f : X ⟶ Y` and a point `x : X.toTopCat`, stating that `f` is smooth at
+`X`.
+-/
+class Scheme.IsSmoothAt {X Y : Scheme} (f : X ⟶ Y) (x : X.toTopCat) : Type _ where
+/- A ring whose prime spectrum is isomorphic to some open nbd of `x`. --/
+  R : CommRingCat
+/- A ring whose prime spectrum is isomorphic to some open subset of `Y` --/
```

Comment:
I think we should instead use the "commutative algebra <-> AG" transport machinery to set this up. In particular you can use https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.html#AlgebraicGeometry.affineLocally to produce a property of morphisms of schemes from a property on ring homomorphisms.

Also for this you can have a look at how it is done for morphisms locally of finite presentation, e.g. https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/Morphisms/FinitePresentation.html#AlgebraicGeometry.LocallyOfFinitePresentation.

Note that showing https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/LocalProperties.html#RingHom.PropertyIsLocal for smooth ring maps is quite tricky and in the works by @erdOne, but it is still possible (and useful!) to make the definition and add a bit of API.